### PR TITLE
NCSDK-28259: Script for generating encryption artifacts and manifest template for encryption

### PR DESCRIPTION
**Pull request by ahasztag**

### CODE REVIEW

File: ncs/app_envelope_encrypted.yaml.jinja2 [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+          - suit-send-record-failure
+          - suit-send-sysinfo-success
+          - suit-send-sysinfo-failure
+    suit-invoke:
+    - suit-directive-set-component-index: 0
+    - suit-directive-invoke:
+      - suit-send-record-failure
+    suit-install:
+    - suit-directive-set-component-index: 1
+    - suit-directive-override-parameters:
```
Comment (kszromek-nordic) [inline ncs/app_envelope_encrypted.yaml.jinja2:73]: This values are already assigned in shared-sequence so probably lines 73-78 are obsolete. I will double check during tests.

File: ncs/encrypt_script.py [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+    SHA_512 = "sha-512"
+    SHAKE128 = "shake128"
+    SHAKE256 = "shake256"
+
+    def __str__(self):
+        return self.value
+
+
+class EncryptionKMSBackends(Enum):
+    """kms backends."""
```
Comment (kszromek-nordic) [inline ncs/encrypt_script.py:68]: NIT: Lowercase letter at beggining of docstring.

File: ncs/encrypt_script.py [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+        SuitDigestAlgorithms.SHA_256.value: hashes.SHA256(),
+        SuitDigestAlgorithms.SHAKE128.value: hashes.SHAKE128(16),
+        SuitDigestAlgorithms.SHA_384.value: hashes.SHA384(),
+        SuitDigestAlgorithms.SHA_512.value: hashes.SHA512(),
+        SuitDigestAlgorithms.SHAKE256.value: hashes.SHAKE256(32),
+    }
+
+    def __init__(self, hash_name: str):
+        """Initialize object."""
+        if hash_name not in self._hash_func.keys():
```
Comment (kszromek-nordic) [inline ncs/encrypt_script.py:97]: No need to change but You can also write `if hash_name not in self._hash_func:` with same effect :)

File: suit_generator/suit/manifest.py [hunk 4]
```diff
@@ -205,6 +205,11 @@ def from_obj(cls, obj: dict) -> SuitUint:
 
             binary_data = SuitEnvelopeTagged.return_processed_binary_data(obj["envelope"])
             return super().from_obj(len(binary_data))
+        elif "file_direct" in obj.keys():
```
Comment (kszromek-nordic) [inline suit_generator/suit/manifest.py:208]: Comment - not issue: `file_direct` is allright. As an alternative, `file_value` might be used.

File: suit_generator/suit/security.py [hunk 5]
```diff
@@ -367,7 +389,55 @@ class CoseEncrypt(SuitTupleNamed):
...
+        }
+    )
+
+
+class SuitEncryptionInfoExt(SuitBstr):
+    """Representation of SUIT encryption info ext."""
+
+    @classmethod
+    def to_obj(self) -> dict:
+        raise ValueError(f"Encryption info should be expanded to full struncture by to_obj method")
```
Comment (kszromek-nordic) [inline suit_generator/suit/security.py:416]: - f-string is not needed
- `struncture` -> `structure`

**Review by kszromek-nordic — APPROVED** [review]
No functional issues have been found. There are only minor comments. Some classes and public methods lack docstrings. Approved anyway.

File: suit_generator/suit/security.py [hunk 6]
```diff
@@ -367,7 +389,55 @@ class CoseEncrypt(SuitTupleNamed):
...
+
+    @classmethod
+    def from_obj(cls, obj: dict) -> SuitBstr:
+        """Restore SUIT representation from passed object."""
+        if not isinstance(obj, dict):
+            raise ValueError(f"Expected dict, received: {obj}")
+        enc_info_bytes = b""
+        if "raw" in obj.keys():
+            enc_info_bytes = obj["raw"]  # TODO: check this
```
Comment (tomchy) [inline suit_generator/suit/security.py:425]: Should not that convert HEX string to bytes object?

File: ncs/encrypt_script.py [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+            raise ValueError(f"Unsupported hash algorithm: {hash_name}")
+        self._hash_name = hash_name
+
+    def generate_digest_for_plain_text(self, plaintext_file_path: Path, output_directory: Path):
```
Comment (tomchy) [inline ncs/encrypt_script.py:101]: Since this method generates two files - I'd include the `size` in the method name too.

**Review by tomchy — CHANGES_REQUESTED** [review]

File: ncs/app_envelope_encrypted.yaml.jinja2 (hunk 1 above)
Comment (tomchy) [reply ncs/app_envelope_encrypted.yaml.jinja2]: Manifests reworked.

File: ncs/app_envelope_encrypted.yaml.jinja2 [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+            suit-parameter-image-size:
+              file_direct: {{ application['encryption_artifacts_dir'] }}/plain_text_size.txt
+        - suit-condition-image-match:
+          - suit-send-record-success
+          - suit-send-record-failure
+          - suit-send-sysinfo-success
+          - suit-send-sysinfo-failure
+      - - suit-directive-override-parameters:
+            suit-parameter-image-size:
+              file: {{ application['encryption_artifacts_dir'] }}/encrypted_content.bin
```
[the remark this answers is not in the document]
Comment (tomchy) [reply ncs/app_envelope_encrypted.yaml.jinja2]: Section removed. The issue was that we need an actual decrypted size to be written before we call the  `release` filter API.

File: ncs/app_envelope_encrypted.yaml.jinja2 [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+          - suit-send-sysinfo-success
+          - suit-send-sysinfo-failure
+    suit-text:
+      suit-digest-algorithm-id: cose-alg-sha-256
+    suit-manifest-component-id:
+    - INSTLD_MFST
+    - RFC4122_UUID:
+        namespace: {{ mpi_application_vendor_name }}
+        name: {{ mpi_application_class_name }}
+  suit-text:
```
[the remark this answers is not in the document]
Comment (tomchy) [reply ncs/app_envelope_encrypted.yaml.jinja2:122]: Updated.

File: ncs/build.py [hunk 10]
```diff
@@ -32,7 +32,11 @@ def read_configurations(configurations):
     """Read configuration stored in the pickled devicetree."""
     data = {}
     for config in configurations:
-        name, binary, edt, kconfig = config.split(",")
+        encryption_artifacts_dir = None
+        if config.count(",") == 4:
```
[the remark this answers is not in the document]
Comment (tomchy) [reply ncs/build.py]: A new logic proposed.

File: suit_generator/suit/security.py (hunk 6 above)
Comment (tomchy) [reply suit_generator/suit/security.py]: Changed to: `bytes.fromhex(obj["raw"])`

File: ncs/encrypt_script.py (hunk 7 above)
Comment (tomchy) [reply ncs/encrypt_script.py]: Method name changed.

**Review by tomchy — APPROVED** [review]

**Review by kszromek-nordic — APPROVED** [review]

**Review by robertstypa — APPROVED** [review]

**Pull request merged**